### ISSUE TITLE
Implement gradle wrapper retry distribution downloading in case of failure

### ISSUE THREAD

**Issue by kalumpiwarra** [gradle/gradle#18124]
### Use case
Currently, gradle wrapper has no retry distribution downloading option. In case of any failures during downloading gradle distribution, it will fail without any excuses.
In my practice sometimes I'm facing an issue that download failed. It could be because of different issues e.g. `Connection has been shutdown: javax.net.ssl.SSLException: java.net.SocketException: Connection reset` or `Downloading from https://services.gradle.org/distributions/gradle-6.9-all.zip failed: timeout`. In that case, I need to restart the job while it will be successful.

### Expected Behavior
I would like to propose an option to retry downloading in case of failure.  The default value will be 0. E.g.
`systemProp.gradle.wrapper.download.retry`

Overall it will be an option to increase stability for distribution downloading.

### Current Behavior
Any connection issue leads to failed distribution downloading and failed to build overall.

### Extra information
I've tried to implement such behavior and in my case, it is working perfectly fine. Around 1 out of 4 builds has such connection issue and retry allows downloading distribution without any issue (relatively to my infrastructure).

**Comment by mmoayyed** [comment]
Seeing a lot of CI failures on GitHub Actions, recently, where it seems like Gradle wrapper cannot download Gradle:

```
2021-11-02T14:31:22.9392343Z Downloading https://services.gradle.org/distributions/gradle-7.2-bin.zip
.
. (Note the time difference)
.
2021-11-02T14:42:00.2593701Z ##[error]The operation was canceled.
```

Having a retry option would be very helpful.

**Comment by KevinKelchen** [comment]
In case anyone finds themselves running into this issue, I wanted to share a workaround.


## Overview

I've created a Node.js script that downloads Gradle with retry attempts and backoff time. Have it run as a step in your build pipeline before the Gradle build step. This will provide the desired version of Gradle for the Gradle build step ahead of time and in a more fail-resistent manner for random failures.


## Algorithm

The number of attempts to make for downloads is set to `5`. The initial backoff time is set to `5 seconds`. When a failure occurs, it waits the duration of the backoff time before trying again. For the next failure that occurs, the backoff time last used is doubled.

If all the attempts failed given this configuration, the timeline should look something like:
  - Attempt 1 failed
  - Wait 5 seconds
  - Attempt 2 failed
  - Wait 10 seconds
  - Attempt 3 failed
  - Wait 20 seconds
  - Attempt 4 failed
  - Wait 40 seconds
  - Attempt 5 failed

These values and the strategy can be adjusted; they were just a "gut feel." It's been observed to work for us without issue during the past month as of writing. You can at least go with this and see how it holds up or adjust as needed.


## Code

Invoke the Node.js script as a step in your build pipeline before the Gradle build step.

The first parameter to `node` is the path to the Node.js script. The second parameter is the path to the `gradlew` file. Adjust the file paths as necessary for your workspace.

```bash
node "./build/download-gradle.js" "./android/gradlew"
```

File contents of `download-gradle.js`:

```javascript
const pathToGradleWrapper = process.argv[2];
const { promisify } = require("util");
const exec = promisify(require("child_process").exec);

// `process` error handlers inspired by https://blog.heroku.com/best-practices-nodejs-errors
process.on("uncaughtException", (err) => {
  logFormatted(`Uncaught Exception: ${err.message}`);
  process.exit(1);
});

process.on("unhandledRejection", (reason) => {
  logFormatted(`Unhandled rejection. reason: ${reason}`);
  process.exit(1);
});

const main = async () => {
  await attemptGradleDownload();
};

const logFormatted = (message) => console.log(`${message}\n\n`);

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

const attemptGradleDownload = async () => {
  const attemptsToMake = 5;
  const initialBackoffTime = 5000;

  logFormatted(`Attempts to make: ${attemptsToMake}`);
  logFormatted(`Initial backoff time: ${initialBackoffTime}ms`);

  let currentBackoffTime = initialBackoffTime;

  for (let attempt = 1; attempt <= attemptsToMake; attempt++) {
    logFormatted(`Attempt ${attempt}`);

    try {
      // Run a Gradle command that is relatively inert but will cause
      // the version of Gradle we're using to be downloaded if needed.
      const runGradlePromise = exec(`${pathToGradleWrapper} --version`);

      runGradlePromise.child.stdout.pipe(process.stdout);

      await runGradlePromise;
    } catch (e) {
      logFormatted(`Attempt ${attempt} failed. Response: ${e}`);

      if (attempt < attemptsToMake) {
        logFormatted(`Waiting ${currentBackoffTime}ms until next attempt.`);

        await sleep(currentBackoffTime);
      }

      currentBackoffTime = currentBackoffTime * 2;

      continue;
    }

    logFormatted(`Gradle command succeeded on attempt ${attempt}`);

    break;
  }
};

(async () => {
  await main();
})();
```


## In Conclusion

Hope this helps for anyone who can run Node.js scripts in their build pipeline! 😀
